packages/mds-policy-service/service/provider.ts

```diff
@@ -14,12 +14,71 @@
-import { PolicyService } from '../@types'
+import { GeographyServiceClient } from '@mds-core/mds-geography-service'
+import logger from '@mds-core/mds-logger'
+import { ProcessController, ServiceException, ServiceProvider, ServiceResult } from '@mds-core/mds-service-helpers'
+import { DependencyMissingError } from '@mds-core/mds-utils'
+import { PolicyMetadataDomainModel, PolicyService, ReadPolicyQueryParams } from '../@types'
 import { PolicyRepository } from '../repository'
+import { validatePolicyDomainModel, validatePolicyMetadataDomainModel } from './validators'
+
+const serviceToPolicyRepository = async <T>(method: string, exec: () => Promise<T>) => {
+  try {
+    return ServiceResult(await exec())
+  } catch (error) {
+    const exception = ServiceException(`Error Policy:${method}`, error)
+    logger.error(`mds-policy-service::${method} error`, { exception, error })
+    return exception
+  }
+}
 
 export const PolicyServiceProvider: ServiceProvider<PolicyService> & ProcessController = {
   start: PolicyRepository.initialize,
   stop: PolicyRepository.shutdown,
-  name: async () => ServiceResult('mds-policy-service')
+  name: async () => ServiceResult('mds-policy-service'),
+  writePolicy: policy =>
+    serviceToPolicyRepository('writePolicy', () => PolicyRepository.writePolicy(validatePolicyDomainModel(policy))),
+  readPolicies: params => serviceToPolicyRepository('readPolicies', () => PolicyRepository.readPolicies(params)),
+  readActivePolicies: timestamp =>
+    serviceToPolicyRepository('readActivePolicies', () => PolicyRepository.readActivePolicies(timestamp)),
+  deletePolicy: policy_id => serviceToPolicyRepository('deletePolicy', () => PolicyRepository.deletePolicy(policy_id)),
+  editPolicy: policy =>
+    serviceToPolicyRepository('editPolicy', () => PolicyRepository.editPolicy(validatePolicyDomainModel(policy))),
+  readPolicy: policy_id => serviceToPolicyRepository('readPolicy', () => PolicyRepository.readPolicy(policy_id)),
+  readSinglePolicyMetadata: policy_id =>
+    serviceToPolicyRepository('readSinglePolicyMetadata', () => PolicyRepository.readSinglePolicyMetadata(policy_id)),
+  updatePolicyMetadata: policy_metadata =>
+    serviceToPolicyRepository('updatePolicyMetadata', () =>
+      PolicyRepository.updatePolicyMetadata(validatePolicyMetadataDomainModel(policy_metadata))
+    ),
+  writePolicyMetadata: policy_metadata =>
+    serviceToPolicyRepository('writePolicyMetadata', () =>
+      PolicyRepository.writePolicyMetadata(validatePolicyMetadataDomainModel(policy_metadata))
+    ),
+  publishPolicy: async (policy_id, publish_date) => {
+    try {
+      const policy = await PolicyRepository.readPolicy(policy_id)
+      const geographies = await GeographyServiceClient.getGeographiesByIds(policy.rules.map(r => r.geographies).flat())
+
+      if (geographies.some(geography => !geography?.publish_date)) {
+        throw new DependencyMissingError(`some geographies not published!`)
+      }
```

Comment:
Does this handle the case that one of the referenced geographies doesn't exist at all?

Reply:
The optional chaining _might_, just wanna be sure :)

Reply:
added a few more tests, factories again!